content/blog/2022-04-19-tracing/index.md

```diff
@@ -0,0 +1,142 @@
++++
+title = "Trace-Based Just-in-Time Type Specialization for Dynamic Languages"
+[extra]
+latex = true
+[[extra.authors]]
+name = "Yulun Yao"
++++
+
+## Introduction
+# TL.DR.
+This paper presents *Trace-Monkey*, a JavaScript language implementation based on tracing Just-In-Time (JIT) compilation. It focuses on identifying frequently executed loop traces at run-time and then compile those traces with dynamic type information. Specifically, the author proposes an algorithm to lazily form trace-trees with alternative paths within the loop, and explains how to generate efficient type-specialized code for the identified traces. The experimental result shows that *Trace-Monkey* is 10x faster compared to previous compilation strategies on some benchmarks.
+
+# Problem
+Dynamic languages, such as JavaScript, Python and Ruby are some of the top programming languages in the recent decade. These languages are accessible to developers thanks to their less restrictive syntax and type systems, easy to deploy on a wide range of platforms since they contain less hardware-specific type information. On the other hand, however, compared to traditional, statically typed languages, namely Java, C++, Ocaml, dynamic languages are much more difficult to be compiled. No hardware would ever support non-typed arithmetic operations, so the typing information needs to be resolved eventually. As types of expressions may vary at runtime in dynamic languages, the typing information is not available to compilers, which very often need to utilize it to do optimizations and generate efficient bytecodes. A seemingly simple yet not-so-simple example for compiler is as follows:
```

Comment:
> dynamic languages are much more difficult to be compiled

Hmm… maybe I'd say they're more difficult to _optimize_? Writing a bad/slow compiler for them doesn't seem hard.